tokio/src/io/unix/pipe.rs

```diff
@@ -0,0 +1,1191 @@
+    /// file; it is left up to the user to make sure it is opened with writing access,
+    /// represents a pipe and is set in non-blocking mode.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use tokio::io::unix::pipe;
+    /// use std::fs::OpenOptions;
+    /// use std::os::unix::fs::{FileTypeExt, OpenOptionsExt};
+    /// # use std::error::Error;
+    ///
+    /// const FIFO_NAME: &str = "path/to/a/fifo";
+    ///
+    /// # async fn dox() -> Result<(), Box<dyn Error>> {
+    /// let file = OpenOptions::new()
+    ///     .write(true)
+    ///     .custom_flags(libc::O_NONBLOCK)
+    ///     .open(FIFO_NAME)?;
+    /// if file.metadata()?.file_type().is_fifo() {
+    ///     let tx = pipe::Sender::from_file_unchecked(file)?;
+    ///     /* use the Sender */
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    pub fn from_file_unchecked(file: File) -> io::Result<Sender> {
+        let raw_fd = file.into_raw_fd();
+        let mio_tx = unsafe { mio_pipe::Sender::from_raw_fd(raw_fd) };
+        Sender::from_mio(mio_tx)
+    }
+
+    /// Waits for any of the requested ready states.
+    ///
+    /// This function can be used to wait for a [`WRITE_CLOSED`] event.
+    ///
+    /// The function may complete without the socket being ready. This is a
+    /// false-positive and attempting an operation will return with
+    /// `io::ErrorKind::WouldBlock`. The function can also return with an empty
+    /// [`Ready`] set, so you should always check the returned value and possibly
+    /// wait again if the requested states are not set.
+    ///
+    /// [`WRITE_CLOSED`]: Ready::WRITE_CLOSED
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is cancel safe. Once a readiness event occurs, the method
+    /// will continue to return immediately until the readiness event is
+    /// consumed by an attempt to read or write that fails with `WouldBlock` or
+    /// `Poll::Pending`.
+    pub async fn ready(&self, interest: Interest) -> io::Result<Ready> {
+        let event = self.io.registration().readiness(interest).await?;
+        Ok(event.ready)
+    }
+
+    /// Waits for the pipe to become writable.
+    ///
+    /// This function is equivalent to `ready(Interest::WRITABLE)` and is usually
+    /// paired with [`try_write()`].
+    ///
+    /// [`try_write()`]: Self::try_write
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use tokio::io::unix::pipe;
+    /// use std::error::Error;
+    /// use std::io;
+    ///
+    /// #[tokio::main]
+    /// async fn main() -> Result<(), Box<dyn Error>> {
```

Comment:
I would prefer to just use `std::io::Error` rather than `Box<dyn Error>` in examples.